drivers/serial/uart_nrfx_uarte.c

```diff
@@ -0,0 +1,629 @@
+{
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	nrf_uarte_int_disable(uarte, NRF_UARTE_INT_ENDRX_MASK);
+}
+
+/** Interrupt driven error enabling function */
+static void uarte_nrfx_irq_err_enable(struct device *dev)
+{
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	nrf_uarte_int_enable(uarte, NRF_UARTE_INT_ERROR_MASK);
+}
+
+/** Interrupt driven error disabling function */
+static void uarte_nrfx_irq_err_disable(struct device *dev)
+{
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	nrf_uarte_int_disable(uarte, NRF_UARTE_INT_ERROR_MASK);
+}
+
+/** Interrupt driven pending status function */
+static int uarte_nrfx_irq_is_pending(struct device *dev)
+{
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	return ((nrf_uarte_int_enable_check(uarte,
+					    NRF_UARTE_INT_ENDTX_MASK) &&
+		 uarte_nrfx_irq_tx_ready_complete(dev))
+		||
+		(nrf_uarte_int_enable_check(uarte,
+					    NRF_UARTE_INT_ENDRX_MASK) &&
+		 uarte_nrfx_irq_rx_ready(dev)));
+}
+
+/** Interrupt driven interrupt update function */
+static int uarte_nrx_irq_update(struct device *dev)
+{
+	return 1;
+}
+
+/** Set the callback function */
+static void uarte_nrfx_irq_callback_set(struct device *dev,
+					uart_irq_callback_t cb)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	data->cb = cb;
+}
+
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+
+static const struct uart_driver_api uart_nrfx_uarte_driver_api = {
+	.poll_in		= uarte_nrfx_poll_in,
+	.poll_out		= uarte_nrfx_poll_out,
+	.err_check		= uarte_nrfx_err_check,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill		= uarte_nrfx_fifo_fill,
+	.fifo_read		= uarte_nrfx_fifo_read,
+	.irq_tx_enable		= uarte_nrfx_irq_tx_enable,
+	.irq_tx_disable		= uarte_nrfx_irq_tx_disable,
+	.irq_tx_ready		= uarte_nrfx_irq_tx_ready_complete,
+	.irq_rx_enable		= uarte_nrfx_irq_rx_enable,
+	.irq_rx_disable		= uarte_nrfx_irq_rx_disable,
+	.irq_tx_complete	= uarte_nrfx_irq_tx_ready_complete,
+	.irq_rx_ready		= uarte_nrfx_irq_rx_ready,
+	.irq_err_enable		= uarte_nrfx_irq_err_enable,
+	.irq_err_disable	= uarte_nrfx_irq_err_disable,
+	.irq_is_pending		= uarte_nrfx_irq_is_pending,
+	.irq_update		= uarte_nrx_irq_update,
+	.irq_callback_set	= uarte_nrfx_irq_callback_set,
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+static int uarte_instance_init(struct device *dev,
+			       const struct uarte_init_config *config,
+			       u8_t interrupts_active)
+{
+	int err;
+	struct device *gpio_dev[GPIO_COUNT];
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	gpio_dev[0] = device_get_binding(CONFIG_GPIO_NRF5_P0_DEV_NAME);
+	__ASSERT(gpio_dev[0],
+		 "UART init failed. Cannot find %s",
+		 CONFIG_GPIO_NRF5_P0_DEV_NAME);
+#if (GPIO_COUNT > 1)
+	gpio_dev[1] = device_get_binding(CONFIG_GPIO_NRF5_P1_DEV_NAME);
+	__ASSERT(gpio_dev[1],
+		 "UART init failed. Cannot find %s",
+		 CONFIG_GPIO_NRF5_P1_DEV_NAME);
+#endif
+	UARTE_PIN_WRITE(gpio_dev, config->pseltxd, 1);
```

Comment:
On second thought, I think it should be done in a different way. Let's get rid of this dependency on gpio devices, which brings only complications here (and currently leads to potential problems, since the gpio devices are initialized POST_KERNEL and we use their API here in a PRE_KERNEL_1 initialization routine), and use nrfx GPIO HAL directly.

Reply:
I agree, I have used nordic HAL functions to configure gpio.